### ISSUE TITLE
why is our KG dumpfile getting so much smaller?

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#314]
<img width="620" alt="screen shot 2019-03-08 at 4 39 13 am" src="https://user-images.githubusercontent.com/5562107/54029089-241a2280-415c-11e9-8f2e-3e632cc45182.png">

**Comment by DeqingQu** [comment]
The dump file is the tarball of the database folder ('graph' folder in our KG).
I checked two KG dump flies: 030819-023708.dump (298 M) and 042618-190247.dump (723 M) and compared the difference between the two folders. After I used the command 'tar -zxvf filename' to unzip the two tarball files, the size of the 030819-023708 folder is 4.4G and the size of the 042618-190247 folder is 1.8G

The files in the tarballs can be divided into two main categories: 'store' files and 'transaction' files. The details of files sizes are screenshot as following:

<img width="571" alt="Screen Shot 2019-03-11 at 9 01 33 AM" src="https://user-images.githubusercontent.com/2810578/54139119-15db3900-43de-11e9-85c8-670cbfb5dca5.png">

The 'transaction' files store the transaction logs of the database and the 'store' files store the nodes, relations and properties of the dababase. The size details of each category is listed as following:

  | 042618-190247.dump | 030819-023708.dump
-- | -- | --
file size | 723M | 298M
size after unzip | 4.4G | 1.8G
size of 'transaction' files | 3.2G | 773M
size of 'store' files | 1.2G | 1G

So the main reason for our KG dumpfiles is getting smaller is the latest KG has less transacation logs. The transaction log retation policy of the KG is set to 1 day. The transaction logs might be cleaned up when the latest KG is generated.

**Comment by DeqingQu** [comment]
For the 'store' files, our latest KG is still smaller (only 1G). There is something intertesting I found when I want to figure out why it happens.

https://neo4j.com/developer/kb/understanding-database-growth/
This post explains the mechanism how the database growth.

I did some similar tests on an empty Neo4j database.
Step1: create an empty db
Step2: insert 10,000 nodes to the db
Step3: remove all nodes in the db
Step4: insert 5,000 **different** nodes to the db

  | total size | size of 'transaction' files | size of 'store' files
-- | -- | -- | --
emtpy db | 180K | 20K | 160K
insert 10000 nodes | 2M | 1.2M | 744K
delete all nodes | 3.4M | 2.5M | 896K
insert 5000 **different** nodes | 3.9M | 3.1M | 820K

After all nodes are deleted from the db, the properties of the deleted nodes are not deleted from file system actually. The data is transfered to a file called 'neostore.nodestore.db.id' for re-use. At the end, there are only 5,000 nodes in the db, but the size of 'store' files is bigger than the db with 10,000 nodes.

The size of the dump file is not linearly related to the number of nodes of the db. For our previous dump files, we may have some other insert or delete operations before the KG is finally generated. The overhead of either data written but no longer in use or the list of eligible IDs to re-use may cause the 'store' files to increase. But when we generate the latest KG from scratch, there is not so much overhead and that is why the size of 'store' files are also getting smaller.
